compiler-rt/lib/profile/InstrProfiling.h

```diff
@@ -100,12 +103,6 @@ ValueProfNode *__llvm_profile_begin_vnodes();
 ValueProfNode *__llvm_profile_end_vnodes();
 uint32_t *__llvm_profile_begin_orderfile();
 
-/*!
- * \brief Clear profile counters to zero.
- *
- */
-void __llvm_profile_reset_counters(void);
```

Comment:
I think removing these methods from the InstrProfiling.h header may break downstream users. Instead can you add a wrapper around these methods in the new header? These wrappers can then be guarded by  __LLVM_INSTR_PROFILE_GENERATE .

Reply:
These functions are not actually removed from this file because the new header `instr_prof_interface.h` is included in this file (see https://github.com/llvm/llvm-project/pull/76471/files#diff-4de780ce726d76b7abc9d3353aef95013e7b21e7bda01be8940cc6574fb0b5ffR16). Are there downstream use cases that pull out `compiler-rt/lib/profile/InstrProfiling.h` and use it by itself?

Reply:
They are effectively removed if the frontend does not define __LLVM_INSTR_PROFILE_GENERATE (see my other comment). I would prefer the dependency to be in the other direction, include InstrProfiling.h in the interface `instr_prof_interface.h`.

> Are there downstream use cases that pull out compiler-rt/lib/profile/InstrProfiling.h and use it by itself?

Yes, we have one internal user that I can find.

Reply:
> They are effectively removed if the frontend does not define __LLVM_INSTR_PROFILE_GENERATE (see my other comment).

`__LLVM_INSTR_PROFILE_GENERATE` is defined explicitly in this header `compiler-rt/lib/profile/InstrProfiling.h` before the new header is included (see https://github.com/llvm/llvm-project/pull/76471/files#diff-4de780ce726d76b7abc9d3353aef95013e7b21e7bda01be8940cc6574fb0b5ffR15) so that we do not depend on the frontend for the macro in the compiler-rt context, and we make sure that these three names are always available (regardless of frontend options). 

I am trying to avoid wrappers or duplicating definitions. [This comment](https://discourse.llvm.org/t/pgo-are-the-llvm-profile-functions-stable-c-apis-across-llvm-releases/75832/7?u=qwu_ibm) seem to imply that creating wrapper or duplicating definition is not desirable. However, if avoiding wrapper or duplication is making downstream code unnecessarily complicated, I will create the wrappers instead.

Reply:
> I would prefer the dependency to be in the other direction, include InstrProfiling.h in the interface instr_prof_interface.h.

Hmmm this is also something I am trying to avoid. The reason is that we are trying to not expose PGO internal definitions to the user. The intention is that a user can safely include `instr_prof_interface.h` in their program to use these public PGO APIs, without introducing all the PGO internal symbols to their programs.

Could you help me understand the advantage for `instr_prof_interface.h` to include `InstrProfiling.h`?

Reply:
Ok, if we don't want to expose the internals then the existing approach seems like the simplest one. I think @davidxl's comment in the discourse thread is about ensuring no control flow changes in profile gen and use in hot code paths. This does not necessarily preclude additional wrappers. 

> Update: I think the explicit definition needs to be guarded by an #ifndef.
Yes, I misread the original as `#ifdef` causing some confusion. This should also be `#undef` afterwards. I was thinking about churn for existing users of InstrProfiling.h but it should be a no-op.

Reply:
> Ok, if we don't want to expose the internals then the existing approach seems like the simplest one. I think @davidxl's comment in the discourse thread is about ensuring no control flow changes in profile gen and use in hot code paths. This does not necessarily preclude additional wrappers.
> 
> > Update: I think the explicit definition needs to be guarded by an #ifndef.
> 
> Yes, I misread the original as `#ifdef` causing some confusion. This should also be `#undef` afterwards. I was thinking about churn for existing users of InstrProfiling.h but it should be a no-op.

Ah got it! Thanks so much for the clarification! `InstrProfiling.h` is updated so we are guarding `__LLVM_INSTR_PROFILE_GENERATE`'s macro definition.